File: week1/cash.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>
/*
 * Author: Lionel Ngoubou
 * Purpose: Greedy Algorithm implementation :
 Return the minimum number of coins needed for a change.
 */

//Update : Improved my solution using array and a for loop

// Number of values in my array
const int N = 4;

int main(void)
{

    // Prompt the user to enter a positive float;
    // if he fails to, re-prompt him for a valid input
    float dollars;
    do
    {
        dollars = get_float("How much: ");
    }
    while (dollars < 0);

    // To handle float imprecision, we round the value to 100
    int change = round(dollars * 100);

    int coins = 0;
    int array[] = {25, 10, 5, 1};
 
    // Minimum number of coins at each iteration is given by the quotient
    // The updated change is the remainder of the division
    for (int i = 0; i < N; i++)
    {
        while (change >= array[i])
        {
            coins += change / array[i];
            change %= array[i];
        }
    }

    printf("Number of coins: %i\n", coins);
}

File: week1/cash.py
# i'm implementing cs50 problem set 1 'cash' in python
# this problem is also named the coin change problem

# prompt the user to enter a number
# in python (3), input is a string so i convert it to float
dollars = float(input("How much:"))
# to handle float imprecision i convert it to an integer and round it
change = round(dollars * 100)
# i initialize a ariable that will keep track of 
# the number of times the loop is executed
# this number represents the minimum of coins for change
coins = 0
while change >= 25:
    change = change - 25
    coins = coins + 1
    
while change >= 10:
    change = change - 10
    coins = coins + 1
    
while change >= 5:
    change = change - 5
    coins = coins + 1
    
while change >= 1:
    change = change - 1
    coins = coins + 1    

print("number of coins is " + str(coins))
